[Bug] Open Card browser crash

###### Reproduction Steps
1. Open Card browser

###### Expected Result
Card browser is opened

###### Actual Result
Crash

###### Debug info
AnkiDroid Version = 2.16alpha80 

Android Version = 12 

Manufacturer = samsung 

Model = SM-A225F 

Hardware = mt6769t 

Webview User Agent = Mozilla/5.0 (Linux; Android 12; SM-A225F Build/SP1A.210812.016; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/104.0.5112.97 Mobile Safari/537.36 

ACRA UUID = 06c5c865-3ffa-4b5a-b91e-878fa8709710 

Scheduler = std3 

Crash Reports Enabled = true 

DatabaseV2 Enabled = true

###### Research
*Enter an [x] character to confirm the points below:*

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

Report at https://ankidroid.org/acra/app/1/bug/11543/report/6327fbb0-d3aa-4c46-8384-f4739dfdbe1f

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=2; index=10
	at com.ichi2.anki.CardBrowser.onCollectionLoaded(CardBrowser.kt:29)
	at com.ichi2.anki.AnkiActivity.startLoadingCollection(AnkiActivity.kt:4)
	at com.ichi2.anki.CardBrowser.onCreate(CardBrowser.kt:13)
```

Could it be connected to https://github.com/ankidroid/Anki-Android/commit/c762f1af254e4f1bea1c207a157d5a24c62170ac?

Got it! @pavreh --> :trophy: 

 Reproduction:
1- go to card browser
2- change second column to deck (others may do it), it will appear to work
3- go to deck list
4- go back to card browser
:boom: 

```
08-24 13:01:32.470 11989 11989 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.CardBrowser}: java.lang.ArrayIndexOutOfBoundsException: length=2; index=2
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7839)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: Caused by: java.lang.ArrayIndexOutOfBoundsException: length=2; index=2
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at com.ichi2.anki.CardBrowser.onCollectionLoaded(CardBrowser.kt:616)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at com.ichi2.anki.AnkiActivity.startLoadingCollection(AnkiActivity.kt:373)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at com.ichi2.anki.CardBrowser.onCreate(CardBrowser.kt:495)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:8051)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:8031)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
08-24 13:01:32.470 11989 11989 E AndroidRuntime: 	... 12 more

```

I tested a revert of just that commit and it worked, I was no longer able to reproduce

#12162 is the PR for the revert then, and I'll release a new alpha as soon as it passes CI and gets a review

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt
@@ -570,7 +570,9 @@ open class CardBrowser :
                     mColumn1Index = pos
                     AnkiDroidApp.getSharedPrefs(AnkiDroidApp.getInstance().baseContext).edit()
                         .putInt("cardBrowserColumn1", mColumn1Index).apply()
-                    mCardsAdapter!!.setColumnOne(COLUMN1_KEYS[mColumn1Index])
+                    val fromMap = mCardsAdapter!!.fromMapping
+                    fromMap[0] = COLUMN1_KEYS[mColumn1Index]
+                    mCardsAdapter!!.fromMapping = fromMap
                 }
             }
 
@@ -597,7 +599,9 @@ open class CardBrowser :
                     mColumn2Index = pos
                     AnkiDroidApp.getSharedPrefs(AnkiDroidApp.getInstance().baseContext).edit()
                         .putInt("cardBrowserColumn2", mColumn2Index).apply()
-                    mCardsAdapter!!.setColumnTwo(COLUMN2_KEYS[mColumn2Index])
+                    val fromMap = mCardsAdapter!!.fromMapping
+                    fromMap[1] = COLUMN2_KEYS[mColumn2Index]
+                    mCardsAdapter!!.fromMapping = fromMap
                 }
             }
 
@@ -608,14 +612,12 @@ open class CardBrowser :
         // get the font and font size from the preferences
         val sflRelativeFontSize = preferences.getInt("relativeCardBrowserFontSize", DEFAULT_FONT_SIZE_RATIO)
         val sflCustomFont = preferences.getString("browserEditorFont", "")
+        val columnsContent = arrayOf(COLUMN1_KEYS[mColumn1Index], COLUMN2_KEYS[mColumn2Index])
         // make a new list adapter mapping the data in mCards to column1 and column2 of R.layout.card_item_browser
         mCardsAdapter = MultiColumnListAdapter(
             this,
             R.layout.card_item_browser,
-            COLUMN1_KEYS[mColumn1Index],
-            COLUMN1_KEYS[mColumn2Index],
-            R.id.card_sfld,
-            R.id.card_column2,
+            columnsContent, intArrayOf(R.id.card_sfld, R.id.card_column2),
             sflRelativeFontSize,
             sflCustomFont
         )
@@ -2054,22 +2056,11 @@ open class CardBrowser :
     inner class MultiColumnListAdapter(
         context: Context?,
         private val resource: Int,
-        private var columnOne: Column,
-        private var columnTwo: Column,
-        private val idColumnOne: Int,
-        private val idColumnTwo: Int,
+        private var fromKeys: Array<Column>,
+        private val toIds: IntArray,
         private val fontSizeScalePcent: Int,
         customFont: String?
     ) : BaseAdapter() {
-
-        fun setColumnOne(from: Column) {
-            columnOne = from
-            notifyDataSetChanged()
-        }
-        fun setColumnTwo(from: Column) {
-            columnTwo = from
-            notifyDataSetChanged()
-        }
         private var mOriginalTextSize = -1.0f
         private var mCustomTypeface: Typeface? = null
         private val mInflater: LayoutInflater
@@ -2078,7 +2069,12 @@ open class CardBrowser :
             val v: View
             if (convertView == null) {
                 v = mInflater.inflate(resource, parent, false)
-                v.tag = kotlin.Pair<TextView, TextView>(v.findViewById(idColumnOne), v.findViewById(idColumnTwo))
+                val count = toIds.size
+                val columns = arrayOfNulls<View>(count)
+                for (i in 0 until count) {
+                    columns[i] = v.findViewById(toIds[i])
+                }
+                v.tag = columns
             } else {
                 v = convertView
             }
@@ -2090,14 +2086,12 @@ open class CardBrowser :
         @KotlinCleanup("Unchecked cast")
         private fun bindView(position: Int, v: View) {
             // Draw the content in the columns
-            val columns = v.tag as kotlin.Pair<TextView, TextView>
-            val columnOneView = columns.first
-            val columnTwoView = columns.second
             val card = mCards[position]
-            setFont(columnOneView)
-            columnOneView.text = card.getColumnHeaderText(columnOne)
-            setFont(columnTwoView)
-            columnTwoView.text = card.getColumnHeaderText(columnTwo)
+            (v.tag as Array<*>)
+                .forEachIndexed { i, col ->
+                    setFont(col as TextView) // set font for column
+                    col.text = card.getColumnHeaderText(fromKeys[i]) // set text for column
+                }
             // set card's background color
             val backgroundColor: Int = getColorFromAttr(this@CardBrowser, card.color)
             v.setBackgroundColor(backgroundColor)
@@ -2145,6 +2139,13 @@ open class CardBrowser :
                 v.typeface = mCustomTypeface
             }
         }
+
+        var fromMapping: Array<Column>
+            get() = fromKeys
+            set(from) {
+                fromKeys = from
+                notifyDataSetChanged()
+            }
 
         override fun getCount(): Int {
             return cardCount

